src/field/base/OdeArray.hh

```diff
@@ -0,0 +1,197 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file OdePointers.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Array.hh"
+#include "base/Range.hh"
+#include "base/Types.hh"
+
+#include <cmath>
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * A state array of the equation of motion: pos = v_[0-2], mom = v_[3-5]
+ *
+ * XXX TODO: extend this to include time and spin or to template with a
+ * variable ode_dim and array type
+ */
+
+class OdeArray
+{
+    static constexpr size_type ode_dim = 6;
+    using ode_type                     = Array<real_type, ode_dim>;
+
+  private:
+    ode_type v_;
+
+  public:
+    // Constructors
+    CELER_FUNCTION
+    OdeArray()
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = 0;
+    }
+
+    CELER_FUNCTION
+    OdeArray(const real_type c)
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = c;
+    }
+
+    CELER_FUNCTION
+    OdeArray(const ode_type v)
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = v[i];
+    }
+
+    // Copy constructor
+    CELER_FUNCTION
+    OdeArray(OdeArray const& v)
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = v[i];
+    }
+
+    // Assignment operator
+    CELER_FUNCTION
+    OdeArray operator=(OdeArray const& v)
+    {
+        for (auto i : range(ode_dim))
+            v_[i] = v[i];
+        return *this;
+    }
+
+    // Access by an index
+    CELER_FUNCTION
+    real_type& operator[](const size_type i)
+    {
+        CELER_ASSERT(i < ode_dim);
+        return v_[i];
+    }
+
+    CELER_FUNCTION
+    real_type const& operator[](const size_type i) const
+    {
+        CELER_ASSERT(i < ode_dim);
+        return v_[i];
+    }
+
+    CELER_FUNCTION
+    ode_type get() const { return v_; }
+
+    // Inplace binary operators
+#define INPLACE_BINARY_OP(OPERATOR)                  \
+    CELER_FUNCTION                                   \
+    OdeArray& operator OPERATOR(const OdeArray& rhs) \
+    {                                                \
+        for (auto i : range(ode_dim))                \
+            v_[i] OPERATOR rhs[i];                   \
+        return *this;                                \
+    }                                                \
+    CELER_FUNCTION                                   \
+    OdeArray& operator OPERATOR(const real_type& c)  \
+    {                                                \
+        for (auto i : range(ode_dim))                \
+            v_[i] OPERATOR c;                        \
+        return *this;                                \
+    }
+    INPLACE_BINARY_OP(+=)
+    INPLACE_BINARY_OP(-=)
+    INPLACE_BINARY_OP(*=)
+    INPLACE_BINARY_OP(/=)
+#undef INPLACE_BINARY_OP
```

Comment:
Do all these operators make sense? Are you really adding the same scalar to both the position and momentum?

Reply:
You are right.  Drop + and - for scalar.  Also did same for ODE_BINARY_OP.